accepted/2021/Globalization/GlobalizationSize.md

```diff
@@ -0,0 +1,81 @@
+
+Since .NET Framework v4.0, the framework mainly depends on the underlying operating system to perform such operations. Before v4.0 the framework was used to carry the globalization data and have code to handle such data (e.g. collations algorithms). Carrying the globalization data was a big burden on the framework because of the maintainability and servicing while the OS's and standards (e.g. Unicode/ICU) already providing all what we need.
+
+In .NET Core, we continued with the same strategy. When running on Windows we depended on NLS Win32 APIs and when running on other platforms (e.g. Linux, Mac OS,...etc.) we depended on the [ICU](https://github.com/unicode-org/icu) library.
+
+In .NET 5.0 we went one step further to [support using ICU library](https://docs.microsoft.com/en-us/dotnet/standard/globalization-localization/globalization-icu) when running on Windows and we made this behavior the default one. Also, we have supported the [ICU app-local](https://docs.microsoft.com/en-us/dotnet/standard/globalization-localization/globalization-icu#app-local-icu) feature which allows applications to use a different ICU version than the global system one if needed.
+
+In .NET Core, we have supported the [Globalization Invariant Mode](https://github.com/dotnet/runtime/blob/main/docs/design/features/globalization-invariant-mode.md) is the mode that the application which doesn't care much about globalization can be configured with mode. The benefit of enabling this mode is avoiding any dependencies on the system or the ICU. The Invariant mode is mainly introduced to support smaller container images (e.g. Alpine Linux image). Another benefit of the Invariant Mode is guaranteeing consistent behavior of the app when running on different OS's or platforms.
+
+### Concerned Scenarios
+
+Although globalization support is working fine for desktop and server applications, it was concerning in other platforms mainly because it required installing ICU library. The ICU module sizes are concerning and wouldn't be acceptable especially when running on mobile platforms.
+
+```
+ICU module sizes from MS-ICU Linux x64 package:
+
+libicudata.so.68.2.0.6    28,320,992
+libicui18n.so.68.2.0.6    4,220,936
+libicuuc.so.68.2.0.6      2,479,488
+```
+
+- **Mobile Platforms**:
+  - **Android**: Although Android OS comes with the ICU libraries which .NET can use, the OEMs can choose to not include such libraries in their device images.
+  - **iOS/tvOS**: don't come with an ICU library that .NET can use. .NET had to provide an ICU package to be used in such platforms.
+- **Web assembly**: The clients have to be a small size and need to run inside the browser which will not allow accessing any libraries outside the browser. That means WebAssembly clients need to include an ICU package for globalization support.
+- Users of the **Alpine Linux containers** which enable `Globalization Invariant Mode` ask for better globalization support without increasing the image size much.
```

Comment:
What about self-contained console apps, are they fine to bundle full ICU?

Reply:
It depends on the app scenario. If doesn't need Globalization features, then can enable Invariant mode. If it does need the globalization features, then need to bundle the ICU package. This is like what we have today.

Reply:
> then need to bundle the ICU package

Right, which makes even trivial self-contained Console.WriteLine app over 30MB bigger

Reply:
Why you think trivial self-contained Console.WriteLine app will need globalization in the first place?

Reply:
Because it's the default and we don't make it anyway easy for developers to move away from it.

Reply:
What would you like to see happen to make this easy?

Reply:
I can see several things we could do better
- tell the developers which APIs or code patterns in their applications are incompatible fully or partially with invariant mode (different calendars, currency names access, string normalization, etc).
- Write code fixers/suggestions to use Invariant culture or prefer APIs which are not culture-specific more aggressively.
- Consider making InvarintMode default mode with an opt-out for some project types (e.g. console apps)